[quant] Add fp32/fp16 zero_point support for CPU fakeQuant

b-koopman opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65055

Differential Revision: [D30975238](https://our.internmc.facebook.com/intern/diff/D30975238)

b-koopman committed:
[quant] Add fp32/fp16 zero_point support for CPU fakeQuant

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975238).

jingsh commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2344,37 +2344,81 @@ void fake_quantize_learnable_tensor_grad_kernel_cpu(
+      // write mask
+      cpu_kernel(iter_mask, [=](SelfType self, float scale, scalar_t zero_point) -> bool {
+        float inv_scale = 1.0f / scale;
+        const auto qval = std::lrintf(zero_point + (self * inv_scale));
```
why not use `std::nearbyintf()` to be consistent?

b-koopman replied:
This is mirroring numerics from `affine_quantizer_base.cpp`,  so keeping std::lrintf for consistency with the quantization implementation it is fakeQuant'ing.

```
/*
* Quantize value based on the following equation
* Xq = Round(Xf * inv_scale + zero_point)
* where zero_point is in float.
*
* Note: For the case of embedding quantization we will set zero_point
* to (-Xmin/scale), where Xmin is the min value in input tensor row.
*/
int quantize_val_float_qparams(float scale, float zero_point, float value, int qmin, int qmax) {
  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
  int qvalue;

  float inv_scale = scale == 0 ? 1.0f : 1.0f / scale;
  // NOLINTNEXTLINE(cppcoreguidelines-narrowing-conversions,bugprone-narrowing-conversions)
  qvalue = lrintf(value * inv_scale + zero_point);
  qvalue = std::max(qmin, std::min(qvalue, qmax));
  return qvalue;
}
```

jingsh commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2344,37 +2344,81 @@ void fake_quantize_learnable_tensor_grad_kernel_cpu(
+
+  if(at::isFloatingType(zero_point_dtype)){
+    // When zero_point is float, quantize mirroring affine quantizer equation
+    // Xq = Round(Xf * inv_scale + zero_point)
```
So in the end, the `Xq` is also floating number? should it be cast to int?

b-koopman replied:
Yes, this is done at L2364 and L2374 via `std::lrintf`

b-koopman replied:
Can you clarify with a code-pointer where you would expect the int cast'ing to happen?  I think this should be handled with `std::lrintf`, then when "fake" de-quantizing, should be a float type again.

jingsh replied:
oh, I was referring to the fake quantized `x` (`xq` here), normally aren't we using the int type as the quantized value? 

But I think it should be good here, as the kernel fake quant then dequant here already

supriyar commented on test/quantization/core/test_workflow_ops.py:
```diff
@@ -804,31 +808,42 @@ def test_backward_per_channel(self, device, X):
             scale, zero_point = obs.calculate_qparams()
             # TODO(future PR): fix the wrong dtype in obs.calculate_qparams and remove the cast
-            zero_point = zero_point.to(torch.int32)
+            zero_point = zero_point.to(torch.int)
```
why aren't we using the `zero_point_type` here?

b-koopman replied:
Fixed, thanks! I think this was an issue when merging.

supriyar commented on test/quantization/core/test_workflow_ops.py:
```diff
@@ -693,15 +693,19 @@ def test_forward_per_channel(self, device, X):
             scale, zero_point = obs.calculate_qparams()
             # TODO(future PR): fix the wrong dtype in obs.calculate_qparams and remove the cast
-            zero_point = zero_point.to(torch.int32)
+            zero_point = zero_point.to(zero_point_type)
```
For any of these tests, do we need to also update the reference implementation to have the same numerics as the op for floating point zero point?

b-koopman replied:
The reference implementation should already match the numerics for floating point zero_point. In the reference impl, at `common_quantized.py:185`, rounding happens like `round(X * 1/scale + zero_point)`, similar to this floating point zero_point implementation in the op.

Somewhat related, I would have expected the reference implementation to match the existing integer op implementation of `static_cast<int64_t>(zero_point + std::nearbyint(self * inv_scale))`.  This might be a good follow-up issue to make sure the reference implementation matches the numerics for existing int zero_point op implementation.

supriyar replied:
Thanks, would be great if we can track the follow-up as a TODO in the test.

b-koopman committed:
Update on "[quant] Add fp32/fp16 zero_point support for CPU fakeQuant"


Differential Revision: [D30975238](https://our.internmc.facebook.com/intern/diff/D30975238)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant] Add fp32/fp16 zero_point support for CPU fakeQuant"


Differential Revision: [D30975238](https://our.internmc.facebook.com/intern/diff/D30975238)

[ghstack-poisoned]
